src/path.cr

```diff
@@ -0,0 +1,941 @@
+
+              io << base_relative
+              io << separators[0] unless base_relative.empty?
+              io << name_relative
+            end
+          end
+        end
+      else
+        expanded = base.join(name)
+      end
+    else
+      expanded = name
+    end
+
+    expanded = new_instance(expanded) unless expanded.is_a?(Path)
+    expanded.normalize(remove_final_separator: false)
+  end
+
+  # Appends the given *parts* to this path and returns the joined path.
+  #
+  # ```
+  # Path{"foo"}.join("bar", "baz")       # => Path{"foo/bar/baz"}
+  # Path{"foo/"}.join("/bar/", "/baz")   # => Path{"foo/bar/baz"}
+  # Path{"/foo/"}.join("/bar/", "/baz/") # => Path{"/foo/bar/baz/"}
+  # ```
+  def join(*parts) : Path
+    join parts
+  end
+
+  # Appends the given *parts* to this path and returns the joined path.
+  #
+  # ```
+  # Path{"foo"}.join("bar", "baz")           # => Path{"foo/bar/baz"}
+  # Path{"foo/"}.join(Path{"/bar/", "/baz"}) # => Path{"foo/bar/baz"}
+  # Path{"/foo/"}.join("/bar/", "/baz/")     # => Path{"/foo/bar/baz/"}
+  # ```
+  #
+  # Non-matching paths are implicitly converted to this path's kind.
+  #
+  # ```
+  # Path.posix("foo/bar").join(Path.windows("baz\baq"))  # => Path.posix("foo/bar/baz/baq")
+  # Path.windows("foo\\bar").join(Path.posix("baz/baq")) # => Path.posix("foo\\bar\\baz/baq")
+  # ```
+  def join(parts : Enumerable) : Path
+    new_name = String.build do |str|
+      str << @name
+      last_ended_with_separator = ends_with_separator?
+
+      parts.each_with_index do |part, index|
+        case part
+        when Path
+          # Every POSIX path is also a valid Windows path, so we only need to
+          # convert the other way around (see `#to_windows`, `#to_posix`).
+          part = part.to_posix if posix? && part.windows?
+
+          part = part.@name
+        else
+          part = part.to_s
+          part.check_no_null_byte
+        end
+
+        if part.empty?
+          if index == parts.size - 1
+            str << separators[0] unless last_ended_with_separator
+            last_ended_with_separator = true
+          else
+            last_ended_with_separator = false
+          end
+
+          next
+        end
+
+        byte_start = 0
+        byte_count = part.bytesize
+
+        case {starts_with_separator?(part), last_ended_with_separator}
+        when {true, true}
+          byte_start += 1
+          byte_count -= 1
+        when {false, false}
+          str << separators[0] unless str.bytesize == 0
+        end
+
+        last_ended_with_separator = ends_with_separator?(part)
+
+        str.write part.unsafe_byte_slice(byte_start, byte_count)
+      end
+    end
+
+    new_instance new_name
+  end
+
+  # Appends the given *part* to this path and returns the joined path.
+  #
+  # ```
+  # Path{"foo"} / "bar" / "baz"     # => Path{"foo/bar/baz"}
+  # Path{"foo/"} / Path{"/bar/baz"} # => Path{"foo/bar/baz"}
+  # ```
+  def /(part : Path | String) : Path
```

Comment:
I see the nice syntax rendering for those used to Unix syntax with slashes `/`.
However `Path / Path` for me means "`Path `divided by `Path`". Plus, the syntax doesn't render as good as for strings.
I suggest having `Path#+(path : Path)` for concatenation.

For `String`, `Path#/(part : String)` is fine.

Reply:
Yeah, that's more or less just me playing around :D

Reply:
I think `/` is fine and it's kind of well known, at least I saw it in some other project (chef, vagrant?)

Reply:
Plus `+` isn't really correct, if you do `"foo" + "bar"` you would intuitively get "foobar", not "foo/bar". So the `/` here makes it very clear that it's being concatenated with a directory separator.

Reply:
I agree for strings, but here I was talking about concatenating two `Path` objects.
`Path` is distribution agnostic, `+` too  (i.e. `/` reminds Unix paths separator)

I'm seeing new Windows confused about the `/`.

Reply:
And keep in mind we'll also have the `Path` affected to variables.
Isn't clearer to do `base_path + current_directory` than `base_path / current_directory`?

Reply:
No, I expect `base_path + current_directory` to concatenate the values, not to stick a `/` in the middle of them.

Reply:
Anyway, in Ruby...

```
$ irb
reirb(main):001:0> require "pathname"
=> true
irb(main):002:0> foo = Pathname.new("foo")
=> #<Pathname:foo>
irb(main):003:0> bar = Pathname.new("bar")
=> #<Pathname:bar>
irb(main):004:0> foo + bar
=> #<Pathname:foo/bar>
irb(main):005:0> foo / bar
=> #<Pathname:foo/bar>
```

So `/` is even there in Ruby.

Reply:
Yes, that's where I got the idea ^^

Reply:
`/` is pretty descriptive in terms of `Path` (and already widely used - see Ruby), noone would really thought of this as `Path` *divided* by `Path` since it makes no sense, OTOH `"foo" / "bar" / "baz.cr"` reads insinctively as a path with separators - like intended.